cacher: do not simply popExpiredWatchers when the cacher hasn't dispatched any event

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
If the cacher hasn't seen any event (when `lastProcessedResourceVersion` is zero) and the `bookmarkTimer` has ticked then we shouldn't just `popExpiredWatchers`. This is because the watchers wont' be re-added and will miss future bookmark events when the cacher finally receives an event via the `c.incoming` chan. This PR fixes the issue by re-adding expired watchers back to the queue. However due to the potential size of the queue we should seek for more performant solution in the future (added a TODO).

xref: https://github.com/kubernetes/enhancements/issues/3157
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cici37 commented:
/triage accepted

p0lyn0mial commented:
/test

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -912,8 +912,6 @@ func (c *Cacher) dispatchEvents() {
 			// because we don't provide any guarantees on sending bookmarks.
 			if lastProcessedResourceVersion == 0 {
-				// pop expired watchers in case there has been no update
-				c.bookmarkWatchers.popExpiredWatchers()
```
I see this as potentially problematic.

Imagine a usecase where no object is actually changing for a long time, but we potentially have a lot of watchers registered.
[FWIW - secrets are a good example]

After this change, until we get the first event, we will be accumulating all watchers in `bookmarkWatchers`.
And those currently are not deleted from `bookmarkWatcher` on watches being closed - they are just are cleaned up here and on dispatching a regular events.

So in this case, we will be accumulating stuff in `bookmarkWatchers`, in a pathological case leading to OOM.

We can't just delete this line - we need a mechanism to prevent it.

p0lyn0mial replied:
Oh, I didn't realise that watchers won't be removed from `bookmarkWatcher` when they are closed. 
I think I would change the code to just "pop" closed watchers, would you agree ?

I think that memory footprint will not be huge. Since we are storing pointers to the `cacheWatcher` struct we should be able to store ~8K entires in just 1MB (assuming the size of the struct is ~100 bytes).

Given the above calculation I think we should be fine.
I could provide a new method `c.bookmarkWatchers.popClosedWatchers()` and just call it in line 916.

wojtek-t replied:
> I think I would change the code to just "pop" closed watchers, would you agree ?

I think we need to ensure that we won't be iterating over too many (still active) watchers every second.
So I think it's not as simple as that.

I think that the "time=0" is somewhat special here...
1) For time>0, probably calling `popExpiredWatchers` as it was before is still the right thing to do?
2) If we do that, we're probably left only with those with `time=0`. For these closing stopped seems good enough, but we need to somehow ensure that we won't be iterating over all of the watchers on every loop...

p0lyn0mial replied:
Actually, I've just realised that the cache receives periodic updates from the `etcd` every X seconds.
These periodic updates are reflected in `lastProcessedResourceVersion`.

Since the graduation of `EfficientWatchResumption` every Kube installation supports progress notification from etcd. In addition to that I think that the default notification period is `5 seconds`. 

So, that means we will be accumulating disconnected watcher for `5 s` in the worst case.
Expired watchers are not re-added again to `watchersBuffer`.

I think we should be fine with what we have atm, wdyt?

wojtek-t replied:
Hmm - under the assumption that etcd is configured to really deliver them.
I can imagine a situation of some operator configuring those progress notifiy events to be delivered every 1h or sth like that. in which case it may become problematic...

I think that even we assume that those with time=0 aren't problematic, I would feel more comfortable if we keep calling `popExpiredWatchers` for time>0...

p0lyn0mial replied:
OK, updated the PR as we discussed offline.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1971,3 +1971,54 @@ func BenchmarkCacher_GetList(b *testing.B) {
+		t.Fatalf("unexpected error waiting for the cache to be ready")
+	}
+
+	opts := storage.ListOptions{Predicate: storage.Everything}
```
nit:
```
pred := storage.Everything
pred.AllowWatchBookmarks = true
opts := storage.ListOptions{
  Predicate: pred,
  AllowWatchBookmarks: ...
}
```

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -926,9 +926,21 @@ func (c *Cacher) dispatchEvents() {
+						if watcher.stopped {
+							continue
+						}
+						c.bookmarkWatchers.addWatcher(watcher)
```
how do you feel about introducing a new function that would allow for adding an array of watchers ?

p0lyn0mial replied:
the benefit would be acquiring the lock just once.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1971,3 +1971,54 @@ func BenchmarkCacher_GetList(b *testing.B) {
+	// the default bookmarkTimer is ~1 sec, waiting 2 seconds is enough
+	time.Sleep(2 * time.Second)
+
+	makePod := func(rv uint64) *example.Pod {
```
Hmm - I'm wondering why do we need this.

1. watchcache is initialized from the dummystorage list with return RV=100
2. then we initialize watch, it goes to storage and returns also RV=100
3. so given it is already synchronized, it should return immediately

What am I missing?

wojtek-t replied:
heh... I'm stupid.

I suggest the following:
1) update the comment above (see my other comment)
2) Here I would ideally switch to imitate progress-notify, but unfortunately this isn't easy...
   So let's leave as is just add a comment like
```
// Send an even to ensure that lastProcessedResourceVersion in Cacher will change to non-zero value.
```

p0lyn0mial replied:
:)

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1971,3 +1971,54 @@ func BenchmarkCacher_GetList(b *testing.B) {
+	require.NoError(t, err, "failed to create watch: %v")
+	defer w.Stop()
+
+	// the default bookmarkTimer is ~1 sec, waiting 2 seconds is enough
```
// Ensure that popExpiredWatchers is called to ensure that our watch isn't removed from bookmarkWatchers.
// We do that every ~1s, so waiting 2 seconds seems enough.

wojtek-t reviewed (commented):
OK - just some minor comments.

wojtek-t commented:
/lgtm
/approve

Thanks!

p0lyn0mial commented:
/hold 

a unit test failure looks real.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -926,9 +958,23 @@ func (c *Cacher) dispatchEvents() {
-				// pop expired watchers in case there has been no update
-				c.bookmarkWatchers.popExpiredWatchers()
+				var expiredWatchers []*cacheWatcher
+				for _, watchers := range c.bookmarkWatchers.popExpiredWatchers() {
```
This should be called under `c.Lock`

wojtek-t commented:
/lgtm cancel

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -206,25 +206,32 @@ func newTimeBucketWatchers(clock clock.Clock, bookmarkFrequency time.Duration) *
-	// so buckedID can hold a negative value
-	nextTime, ok := w.nextBookmarkTime(t.clock.Now(), t.bookmarkFrequency)
-	if !ok {
+	wantsBookmark, bucketID := t.calculateNextBookmarkBucketID(w)
```
Also - can you please move the refactoring you did here to a separate PR?

I really prefer having bug fixes sit in their own PRs.

p0lyn0mial committed:
cacher: do not popExpiredWatchers when the cacher hasn't dispatched any event

If the cacher hasn't seen any event (when lastProcessedResourceVersion is zero) and
the bookmarkTimer has ticked then we shouldn't popExpiredWatchers. This is
because the watchers wont' be re-added and will miss future bookmark events when
the cacher finally receives an event via the c.incoming chan.

p0lyn0mial commented:
I rebased this PR on top of https://github.com/kubernetes/kubernetes/pull/117410, PTAL.

p0lyn0mial commented:
/hold cancel

wojtek-t commented:
/lgtm
/approve
